Migrate database to new structure

fabian-emilius opened the pull request:
Migrate database to new structure proposed in https://github.com/ls1intum/thesis-track/issues/118 

*Changes*
- Changed "semester" input to "enrolled at" in application form to prevent that the student's semester gets outdated
- New review application modal to extract review logic out of the application form
  - Show PDFs in iframe instead of having to download it every time
  - Adding comment or advisor in same request like accepting or rejecting the application
- Refactored and improved server code for new database structure
- Sync keycloak roles to db to allow queries for advisors, supervisors or admins
- Migrate old data to new table structure

*Deployment Changes*
I tried to make environment variables more consistent in this PR. This comes with the following changes
- SERVER_HOST and CLIENT_HOST need to start with the protocol now
- Service hosts are handled in docker-compose.prod now and only access data is from environment
- Add MAIL_WORKSPACE_URL that has the slack invite url used in emails
- Renamed IOS_SENDER_MAIL to MAIL_SENDER
- Renamed CHAIR_MEMBER_RECIPIENTS  to MAIL_CHAIR_RECIPIENTS 
- Add SPRING_DATASOURCE_DATABASE that defines the db name

iyannsch commented on server/src/main/java/thesistrack/ls1/entity/ThesisAssessment.java:
```diff
@@ -0,0 +1,53 @@
+    private Thesis thesis;
+
+    @NotNull
+    @Column(name = "summary", nullable = false, length = Integer.MAX_VALUE)
```
MAX_VALUE values are huge objects in the DB. Assuming 3000 chars/page, I do not think that a summary will even be remotely close to 715.000 pages 😮 Please reconsider your `length` parameter

fabian-emilius replied:
I did not really want that admins run into limits, but yeah makes sense to change it to something more reasonable

iyannsch commented on server/src/main/java/thesistrack/ls1/entity/User.java:
```diff
@@ -0,0 +1,114 @@
+    @Column(name = "university_id", nullable = false, length = 100)
+    private String universityId;
+
+    @Column(name = "matriculation_number", length = 100)
```
```suggestion
    @Column(name = "matriculation_number", length = 30)
```

iyannsch commented on server/src/main/java/thesistrack/ls1/entity/User.java:
```diff
@@ -0,0 +1,114 @@
+    private UUID id;
+
+    @NotNull
+    @Column(name = "university_id", nullable = false, length = 100)
```
```suggestion
    @Column(name = "university_id", nullable = false, length = 30)
```

iyannsch commented on server/src/main/java/thesistrack/ls1/entity/User.java:
```diff
@@ -0,0 +1,114 @@
+    @Column(name = "last_name", length = 100)
+    private String lastName;
+
+    @Column(name = "gender", length = 100)
```
```suggestion
    @Column(name = "gender", length = 20)
```

iyannsch commented on server/src/main/java/thesistrack/ls1/entity/User.java:
```diff
@@ -0,0 +1,114 @@
+    @Column(name = "study_program", length = 100)
+    private String studyProgram;
+
+    @Column(name = "projects", length = 2000)
+    private String projects;
+
+    @Column(name = "interests", length = 2000)
+    private String interests;
```
Please reduce length according to feedback in model

iyannsch commented on server/src/main/java/thesistrack/ls1/repository/UserRepository.java:
```diff
@@ -0,0 +1,27 @@
+public interface UserRepository extends JpaRepository<User, UUID> {
+    Optional<User> findByUniversityId(String universityId);
+
+    @Query("SELECT u FROM User u JOIN UserGroup g ON (u.id = g.id.userId) WHERE " +
+            "g.id.group IN (:groups) AND " +
+            "(LOWER(u.firstName) LIKE %:searchQuery% OR " +
+            "LOWER(u.lastName) LIKE %:searchQuery% OR " +
+            "LOWER(u.email) LIKE %:searchQuery% OR " +
+            "LOWER(u.matriculationNumber) LIKE %:searchQuery% OR " +
+            "LOWER(u.universityId) LIKE %:searchQuery%)")
+    Page<User> searchUsers(@Param("searchQuery") String searchQuery, @Param("groups") Set<String> groups, Pageable page);
```
In `ApplicationRepository.searchApplications()` you require the `groups` parameter as first param, here you require it as second. This results in switched arguments and reduces consistency. Please fix either of those 👍

fabian-emilius replied:
Yes makes sense

iyannsch commented on server/src/main/java/thesistrack/ls1/service/ApplicationService.java:
```diff
@@ -0,0 +1,219 @@
+        this.thesisService = thesisService;
+    }
+
+    public Page<Application> getAll(String searchString, ApplicationState[] states, int page, int limit, String sortBy, String sortOrder) {
```
See my comment above regarding the argument order. Here, the `searchString` (aka `query`) is also flipped with the `state`

fabian-emilius replied:
Makes sense

iyannsch commented on server/src/main/java/thesistrack/ls1/service/ApplicationService.java:
```diff
@@ -0,0 +1,219 @@
+        });
+        Application application = new Application();
+
+        student.setUniversityId(payload.getUniversityId());
+        student.setMatriculationNumber(payload.getMatriculationNumber());
+        student.setIsExchangeStudent(payload.getIsExchangeStudent());
+        student.setFirstName(payload.getFirstName());
+        student.setLastName(payload.getLastName());
+        student.setGender(payload.getGender());
+        student.setNationality(payload.getNationality());
+        student.setEmail(payload.getEmail());
+        student.setStudyDegree(payload.getStudyDegree());
+        student.setStudyProgram(payload.getStudyProgram());
+        student.setEnrolledAt(payload.getEnrolledAt());
+        student.setSpecialSkills(payload.getSpecialSkills());
+        student.setInterests(payload.getInterests());
+        student.setProjects(payload.getProjects());
+        student.setResearchAreas(payload.getResearchAreas());
+        student.setFocusTopics(payload.getFocusTopics());
+        student.setUpdatedAt(currentTime);
+
+        student.setExaminationFilename(uploadService.store(examinationReport));
+        student.setCvFilename(uploadService.store(cv));
+
+        if (degreeReport != null && !degreeReport.isEmpty()) {
+            student.setDegreeFilename(uploadService.store(degreeReport));
+        }
+
+        application.setUser(student);
+        application.setThesisTitle(payload.getThesisTitle());
+        application.setMotivation(payload.getMotivation());
+        application.setState(ApplicationState.NOT_ASSESSED);
+        application.setDesiredStartDate(payload.getDesiredStartDate());
+        application.setCreatedAt(currentTime);
+
+        mailingService.sendApplicationCreatedMailToChair(application);
```
If I see this correctly, there is no sanitization in place, is it? Neither in the `ApplicationRepository` nor here are values checked and mostly accepted straight out of the payload. This poses a security risk.. Did I miss a sanitization aspect in your system?

iyannsch replied:
Down below in `accept()` you perform sanitization ☺️👍

fabian-emilius replied:
I have added more request validation now

iyannsch commented on server/src/main/java/thesistrack/ls1/security/JwtAuthConverter.java:
```diff
@@ -18,40 +17,35 @@
-            claimName = properties.getPrincipalAttribute();
-        }
-        return jwt.getClaim(claimName);
+        return new JwtAuthenticationToken(jwt, authorities, jwt.getClaim(config.getUniversityIdJwtAttribute()));
```
```suggestion
        // Using the universityId as JWT's name attribute
        return new JwtAuthenticationToken(jwt, authorities, jwt.getClaim(config.getUniversityIdJwtAttribute()));
```

fabian-emilius replied:
makes sense

iyannsch commented on server/src/main/java/thesistrack/ls1/constants/ThesisRoleName.java:
I think you could create a more holistic `ThesisRole` enum holding all required information instead of have, e.g., a `ThesisRoleName` object. I'd suggest refactoring this (and other `ThesisRoleXYZ` enums) into one big enum. @krusche what do you think of this?

fabian-emilius replied:
ThesisRoleName is the enum of possible roles and ThesisRole is the database entity that maps the roles to users

iyannsch commented on server/src/main/java/thesistrack/ls1/service/MailingService.java:
```diff
@@ -10,217 +10,242 @@
-    public String getMailTemplate(final String filename) {
-        return storageService.readFromFile(rootLocation, filename + ".html");
+        if (chairMemberRecipientsList != null && !chairMemberRecipientsList.isEmpty()) {
+            this.chairMemberRecipientsList = Arrays.asList(chairMemberRecipientsList.split(";"));
```
Couldn't you use your `StringToArrayConverter` here? ;)

fabian-emilius replied:
Mmh it has a different separator and still exist because of legacy code. The StringToArrayConverter is mainly for request parameters and not configuration but I will consider refactoring this later

iyannsch commented on client/src/requests/responses/user.ts:
As `IUser` is strictly larger than `ILightUser`, you could use interface inheritance here and reduce redundancy throughout the models. Please find more information [here](https://www.codecademy.com/learn/learn-typescript/modules/learn-typescript-advanced-object-types/cheatsheet) 
<img width="459" alt="image" src="https://github.com/user-attachments/assets/d05a0c61-eeb4-4cc5-bc6e-41c31ac4a78c">

fabian-emilius replied:
changed. Makes sense

fabian-emilius commented on client/src/components/UploadArea/UploadArea.tsx:
very good catch. For my small testing file it always worked but does not seem to work for bigger files because limitation was 3KB instead of 3MB. I have pushed changes

fabian-emilius commented on client/src/contexts/AuthenticationContext/AuthenticationProvider.tsx:
```diff
@@ -117,7 +119,10 @@ const AuthenticationProvider = (props: PropsWithChildren<IAuthenticationProvider
-      } else if (decodedAccessToken?.exp && decodedAccessToken.exp <= Date.now() / 1000 + refreshTokenFrequency) {
+      } else if (
+        decodedAccessToken?.exp &&
+        decodedAccessToken.exp <= Date.now() / 1000 + refreshTokenFrequency
```
The idea is that the token is already refreshed when it would expire before the next check. I check every refreshTokenFrequency whether the token is expired and it just adds this interval to the check.
I need to divide Date.now() by 1000 because it returns milliseconds and in the JWT token the timestamp is stored in seconds

fabian-emilius commented on client/src/contexts/AuthenticationContext/AuthenticationProvider.tsx:
```diff
@@ -138,19 +143,37 @@ const AuthenticationProvider = (props: PropsWithChildren<IAuthenticationProvider
     if (authenticationTokens?.access_token) {
       const decodedAccessToken = jwtDecode<IDecodedAccessToken>(authenticationTokens.access_token)
 
-      setUser({
-        first_name: decodedAccessToken.given_name,
-        last_name: decodedAccessToken.family_name,
-        email: decodedAccessToken.email,
-        university_id: decodedAccessToken[GLOBAL_CONFIG.keycloak.university_id_jwt_field] || '',
-        user_id: decodedAccessToken[GLOBAL_CONFIG.keycloak.university_id_jwt_field] || '',
-        roles: decodedAccessToken.resource_access[GLOBAL_CONFIG.keycloak.client_id]?.roles ?? [],
-      })
```
Its still same file but the User object is loaded from the backend now, so basically this logic is run in the backend. It uses POST /user-info to update and get the user info from the database based on the data in the JWT token

fabian-emilius commented on client/src/pages/LegacyApplicationReviewPage/components/LegacyApplicationReviewModal/LegacyApplicationReviewModal.tsx:
```diff
@@ -0,0 +1,339 @@
+              {application.reviewedAt && (
+                <LabeledItem label='Reviewed At' value={application.reviewedAt} />
+              )}
+              {application.comment && (
+                <LabeledItem label='Review Comment' value={application.comment} />
+              )}
```
There was some back and forth with the stakeholders, but the comment field is added again

fabian-emilius commented on client/src/pages/LegacyApplicationReviewPage/components/LegacyApplicationReviewModal/LegacyApplicationReviewModal.tsx:
```diff
@@ -0,0 +1,339 @@
+              {...form.getInputProps('notifyUser', { type: 'checkbox' })}
+            />
+
+            <Group grow>
+              <Button
```
I did changes so cant accept change here but will commit it myself

fabian-emilius commented on client/src/pages/LegacyApplicationReviewPage/components/LegacyApplicationReviewModal/LegacyApplicationReviewModal.tsx:
```diff
@@ -0,0 +1,339 @@
+              >
+                Reject
+              </Button>
+              <Button
```
I did changes so cant accept change here but will commit it myself

fabian-emilius commented on client/src/requests/responses/user.ts:
Yes makes sense

fabian-emilius commented on docker-compose.prod.yml:
```diff
@@ -41,40 +41,31 @@ services:
-      - ENVIRONMENT
-      - SPRING_DATASOURCE_URL
+      - SPRING_JPA_HIBERNATE_DDL_AUTO=update
+      - SPRING_DATASOURCE_URL=jdbc:postgresql://db:5432/${SPRING_DATASOURCE_DATABASE}
```
I dont really see a reason why the database host needs to be configured in the environment variables because the database host is set in the docker compose file so it makes sense to set the same name in the same config and not have different places for the same name

iyannsch replied:
You are correct with the "same place" 👍 I just feel like the port should be configurable in one place and not in `server` and `db`

fabian-emilius replied:
mmh but the port is completely internal so whatever you would put in the environment variable would be used as port and would not make any difference as database cant be accessed from the outside

fabian-emilius replied:
I feel quite strong about the way I did it here

fabian-emilius commented on docker-compose.prod.yml:
```diff
@@ -41,40 +41,31 @@ services:
-      - ENVIRONMENT
-      - SPRING_DATASOURCE_URL
+      - SPRING_JPA_HIBERNATE_DDL_AUTO=update
+      - SPRING_DATASOURCE_URL=jdbc:postgresql://db:5432/${SPRING_DATASOURCE_DATABASE}
```
The database initialization is PostgreSQL code that is not fully compatible with MySQL so this would require larger changes which I would avoid for now

fabian-emilius commented on client/src/requests/payloads/application.ts:
It does not have all fields from user and has some additional ones. I prefer to decouple interfaces like this to not run into issues if you rename fields as its used for something completely different: creating an application <-> representing a user. It would be possible to share some interfaces but I think it would be bad practice

fabian-emilius commented on client/src/utils/blob.ts:
I have changed the file name and especially extension to be more strict in the frontend.

iyannsch replied:
Does your client rely on downloading files other than PDFs?

fabian-emilius commented on server/src/main/java/thesistrack/ls1/controller/ThesisController.java:
```diff
@@ -0,0 +1,132 @@
+
+@Slf4j
+@RestController
+@RequestMapping("/v1/theses")
```
Right now there is only v2 for applications as the only legacy endpoint is to create a application because in the future you will need to be authenticated for that. All other endpoints use v1 at the moment as there are no other legacy endpoints. Do you think all my endpoints should start with v2 except for the one legacy endpoint?

iyannsch replied:
I think all new endpoints / controllers and those relying on "new" DTOs should use a `/v2` path for consistency. Semantically, I'd expect v1 endpoints to be older/deprecated when v2 endpoints are available

fabian-emilius replied:
I see arguements for both sides but you convinced me, so I will change it

fabian-emilius commented on server/src/main/java/thesistrack/ls1/controller/ThesisController.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    @DeleteMapping("/{topicId}")
+    public ResponseEntity<ThesisDto> closeThesis(@PathVariable UUID topicId) {
```
should be thesisId. Good catch!

fabian-emilius commented on server/src/main/java/thesistrack/ls1/controller/ThesisController.java:
```diff
@@ -0,0 +1,132 @@
+        throw new ResponseStatusException(HttpStatus.NOT_IMPLEMENTED, "This feature is not implemented yet");
+    }
+
+    @PostMapping("/{thesisId}/presentations")
+    public ResponseEntity<ThesisDto> createPresentation(@PathVariable UUID thesisId) {
+        throw new ResponseStatusException(HttpStatus.NOT_IMPLEMENTED, "This feature is not implemented yet");
+    }
```
the POST endpoint schedules a presentation with location, time etc and the PUT endpoint uploads the slides of the presentation. Its not implemented yet so this is only my idea of endpoints going forward but still might change

iyannsch replied:
I understand (and like) your rationale ☺️ Have you considered using different paths for this different functionality? I feel like changing only the method (POST, PUT) might be confusing in future

fabian-emilius commented on server/src/main/java/thesistrack/ls1/controller/ThesisController.java:
```diff
@@ -0,0 +1,132 @@
+        throw new ResponseStatusException(HttpStatus.NOT_IMPLEMENTED, "This feature is not implemented yet");
+    }
+
+    @DeleteMapping("/{thesisId}/presentations/{presentationId}")
+    public ResponseEntity<ThesisDto> deletePresentation(@PathVariable UUID thesisId, @PathVariable UUID presentationId) {
+        throw new ResponseStatusException(HttpStatus.NOT_IMPLEMENTED, "This feature is not implemented yet");
+    }
```
This refers to a scheduled presentation and not the uploaded presentation. I see its a bit confusing but will make more sense once code is implemented

fabian-emilius commented on server/src/main/java/thesistrack/ls1/controller/ThesisController.java:
```diff
@@ -0,0 +1,132 @@
+        throw new ResponseStatusException(HttpStatus.NOT_IMPLEMENTED, "This feature is not implemented yet");
+    }
+
+    @PostMapping("/{thesisId}/assessment")
+    public ResponseEntity<ThesisDto> createAssessment(@PathVariable UUID thesisId) {
+        throw new ResponseStatusException(HttpStatus.NOT_IMPLEMENTED, "This feature is not implemented yet");
+    }
+
+    @PostMapping("/{thesisId}/grade")
+    public ResponseEntity<ThesisDto> addGrade(@PathVariable UUID thesisId) {
+        throw new ResponseStatusException(HttpStatus.NOT_IMPLEMENTED, "This feature is not implemented yet");
+    }
```
Right now the idea is that they will be part of the Thesis Entity or ThesisDto Object and not a seperate object

iyannsch replied:
If you offer the functionality to update/create them separately (=seperate objects), I think you should stay consistent and also offer a GETter for them. Once an API object, always an API object. However, this might introduce a significant (network) overhead when updating the ThesisEntity every time.. Why don't you want to have separate SETter for this?

fabian-emilius commented on server/src/main/java/thesistrack/ls1/controller/UserController.java:
```diff
@@ -0,0 +1,71 @@
+    @GetMapping("/{userId}/examination-report")
+    @PreAuthorize("hasAnyRole('admin', 'advisor', 'supervisor')")
+    public ResponseEntity<Resource> getExaminationReport(@PathVariable UUID userId) {
+        return ResponseEntity.ok()
```
If file is not found, it throws an exception which will return an 500 error.

fabian-emilius commented on server/src/main/java/thesistrack/ls1/controller/UserController.java:
```diff
@@ -0,0 +1,71 @@
+    @GetMapping("/{userId}/cv")
+    @PreAuthorize("hasAnyRole('admin', 'advisor', 'supervisor')")
+    public ResponseEntity<Resource> getCV(@PathVariable UUID userId) {
+        return ResponseEntity.ok()
```
If file is not found, it throws an exception which will return an 500 error.

fabian-emilius commented on server/src/main/java/thesistrack/ls1/controller/UserController.java:
```diff
@@ -0,0 +1,71 @@
+    @GetMapping("/{userId}/degree-report")
+    @PreAuthorize("hasAnyRole('admin', 'advisor', 'supervisor')")
+    public ResponseEntity<Resource> getDegreeReport(@PathVariable UUID userId) {
+        return ResponseEntity.ok()
```
If file is not found, it throws an exception which will return an 500 error.

fabian-emilius commented on server/src/main/java/thesistrack/ls1/controller/UserInfoController.java:
UserController is for fetching data about any users. UserInfoController is for updating the authenticated user (Updating database based on JWT data or updating the user profile by the user ). Could be done in one controller but I thought it would be smoother to split this and have the UserController focus only on endpoints about any user which is unrelated to the currently logged in user

iyannsch replied:
This is a design decision and your way of handling the difference between all users and current user is completely fine. Works for me :)

fabian-emilius commented on server/src/main/java/thesistrack/ls1/entity/Application.java:
```diff
@@ -0,0 +1,62 @@
+    private String thesisTitle;
+
+    @NotNull
+    @Column(name = "motivation", nullable = false, length = 2000)
```
Makes sense. I can reduce limit.

fabian-emilius commented on server/src/main/java/thesistrack/ls1/entity/Schedule.java:
I will use this to store cronjob data like: Send a reminder to application reviewer every 3 days to notify them about unreviewed applications

fabian-emilius commented:
@iyannsch Thank you very much for your detailed review of this large PR. In the future I dont plan to do this large PRs anymore. Here it was needed because I changed the complete database architecture which had implications on the whole codebase. There probably would be some smaller changes that could be done in a separate PR but the majority of changes were needed to deliver a working version.

Thanks for the suggestion with the records for DTOs. This definitely makes more sense and I have changed that.

I have also addressed and changed most of your other suggestions and findings. I left a couple of conversations open as you might have a different opinion, so feel free to check out my answers.

iyannsch commented on server/src/main/java/thesistrack/ls1/controller/ApplicationController.java:
```diff
@@ -0,0 +1,121 @@
+                payload.getComment(),
+                payload.getNotifyUser(),
+                payload.getCloseTopic()
+        );
```
Please explain to me where the sanitization now takes place ☺️ Even though you marked this as resolved, I do not see said logic in `ApplicationController.java`

iyannsch commented on server/src/main/java/thesistrack/ls1/constants/ThesisState.java:
```diff
@@ -6,13 +6,13 @@
-    GRADED("Graded"),
-    FINISHED("Finished"),
-    DROPPED_OUT("Dropped out");
+    PROPOSAL("proposal"),
+    WRITING("writing"),
+    SUBMITTED("submitted"),
+    ASSESSED("assessed"),
+    GRADED("graded"),
+    FINISHED("finished"),
+    DROPPED_OUT("dropped_out");
```
In `ApplicationState`, you use UPPER_CASEd names as a good practice. Let's do the same throughout all enums? 
```suggestion
    PROPOSAL("PROPOSAL"),
    WRITING("WRITING"),
    SUBMITTED("SUBMITTED"),
    ASSESSED("ASSESSED"),
    GRADED("GRADED"),
    FINISHED("FINISHED"),
    DROPPED_OUT("DROPPED_OUT");
```

iyannsch commented on server/src/main/java/thesistrack/ls1/dto/UserDto.java:
Can't we use the `LightUserDto` and perform proper inheritance? 👍

fabian-emilius replied:
Records do not support inheritance unfortunately

iyannsch commented on server/src/main/java/thesistrack/ls1/service/UploadService.java:
```diff
@@ -63,7 +68,15 @@ public String store(MultipartFile file) {
     }
 
     public FileSystemResource load(String filename) {
```
Please add a check (like above) to stop the client from inputting `..` values breaking out of the `rootLocation`

fabian-emilius replied:
makes sense. Adds a bit more security in case of bugs

iyannsch reviewed (changes requested):
I added a couple new comments, but nothing too big ☺️ Good job implementing my prior aspects 🔥

Regarding the functionality to upload and - especially - retrieve local files from the upload directory, I'm a bit worried that this is not secure. What if I simply guess the filename of some other student's CV and send a respectively forged GET call to the server? As the original filename is persisted, I could probably also overwrite some other student's file with mine by uploading it under the same name, no? 
We might need to think of some authorization here

fabian-emilius commented:
Thanks for your review @iyannsch . I have changed  the last suggestions except for the database port in docker compose because in my opinion it would be a configuration that has 0 impact on anything.

Regarding the file upload: Right now it is implemented in a way that a user never retrieves a filename. A user calls an endpoint to get a CV of a specific user id but never requests just access to a file. The filename is generated with UploadService.store, which is just a hash of the content, and stored in the db to retrieve it later. The user cant decide the filename and can only overwrite the file if it has the same content (hash is the same)

iyannsch reviewed (approved):
You are right, good idea with the hashed filename 👍 Code LGTM
